### ISSUE TITLE
EmberData | 5.0 Roadmap

### ISSUE THREAD

**Issue by runspired** [emberjs/data#8086]
# 5.0 Roadmap

**Target Date:** Nov/Dec 2022

> **Note** Tasks marked `[-]` have been moved to the 6.0 Roadmap #8460 

## Overview

The **EmberData** Team intends to release a major version of the library around/during the fall of 2022. This issue tracks the work to be done to prepare for that major. It's a non-comprehensive list and some things will likely be added to it as unknowns are discovered over time.

5.0 will ship with an `ember-data` optimized for Embroider out of the box. This should significantly improve build times for all apps (even traditional ember-cli builds), allow ember-data to be chunked, and bring all of the asset compilation optimizations that ember-cli users currently receive to embroider users. Today embroider users ship 2-3x more compressed ember-data code to production than users of traditional ember-cli. This effort is tracked in #8103 

This major version will focus on four critical points of evolving the data story for Ember applications: many of these will be items that will need to be individually RFC'd.

1. removal of all support for Ember classic
2. removal of all promise proxies
3. introduction of a new story for managing network requests
4. introduction of a model-optional default story for presenting data in applications

There's a lot to unpack in those four work items. Let’s dig in:

>  ### 1. removal of all support for Ember classic

Some features can only work in classic mode if we build them using classic APIs. This includes features like decorators which must be built over computed, and base classes (such as Model) which must be built upon EmberObject. Building these APIs in this way introduces unintentional complications that limit what EmberData can do and prevent us from offering better solutions. For instance, the nature of `reopen` and `reopenClass` and `Mixins` are such that even if we were to deprecate classic class syntax, any custom decorators we write for schema information may fall victim to what these methods can do to a Model.

Historically, significant portions of internal code, typically the densest and most complex portions, were developed to defend against and manage the consequences of observers, computed chains and "mandatory setter" wreaking havoc by accessing store state before it was finalized. This did not become simpler with Octane, but harder, because EmberData now has to juggle two fundamentally opposed forms of change notification. Bug reports around change notification are the most common reports we receive. Removing support for observers and chains will significantly simplify some portions of the codebase allowing us to offer more advanced feature sets at lower performance costs and with more reliable change notification semantics. It would even allow for some portions of the library to be reimagined to use `tracked` properties directly, something we have to this point been unable to do due to the consequences of observers. While the observer problem was made significantly better with the deprecation and removal of sync-observers in ember 4.x, there's still much more for us to achieve.

While Ember is intending to deprecate support for classic, we feel EmberData should make the first move here due to the simple fact that we cannot support classic without building with classic APIs, and were we to not remove this support in advance then the resulting Ember deprecations would be unresolvable for our users.

What does this mean in practice?

- Users must convert Adapter/Serializer/Model usage to native class syntax
- We will remove EmberObject (and its APIs) from the prototype chain of all publicly extendable classes
- Adapter Errors will no longer extend Ember Error
- attr/belongsTo/hasMany would only work for native syntax once the deprecation lifecycle is complete
- all array-like APIs will be removed in favor of Arrays or custom iterables. Therein, calling `A()` on any array or iterable returned by EmberData would be a "no-op" operation.
- where possible the detection of a computed or observer chain interacting with an EmberData provided class will print a deprecation. In 5.0 any said interaction would either fail (with a nice assert in dev if detectable) or result in unsupported and undefined behaviors.

>  ### 2. removal of all promise proxies

Several RFCs have already begun this work. All promise proxies would be eliminated in favor of awaiting the async value before usage. Promise state flags would be provided via a utility such as the awesome `ember-promise-helpers` library by @fivetanley. We feel such a utility should be given a home core to the ecosystem in a way that all libraries could utilize it, and would likely bring it in as an official package. Promise proxies currently make it nearly impossible to use `async await` within the ember-data codebase when also using typescript and properly typing a method. While we could eventually decide to sometimes provide a custom async value, simply returning promises as the default for async is better for tooling, typescript, and expectations.

>  ### 3. introduction of a new story for managing network requests

Adapters and Serializers got us far, but simpler approaches with greater flexibility, maintainability, and customizability have been proved out. Watch for a RequestManager RFC. 5.0 would not itself remove support for adapters and serializers, as we can easily support them as a custom plugin for the new system for some time, but it would move their usage into a "legacy" mode and remove them from the default story.

>  ### 4. introduction of a model-optional default story for presenting data in applications

ember-m3, ember-data-model-fragments, ember-data-changetracker, ember-data-storefront and so many other libraries along the way have each tried to solve how to model more complex data and interactions than what `@ember-data/model` has provided out of the box. ember-m3 was a proving ground that we can do better by default, and it's learnings have led to internal refactoring, new public APIs, and new RFCs to allow us to present data in a whole new approach that fits a larger number of APIs more naturally, is more adept at adapting to new situations, and scales more naturally as applications and teams grow. Critically, these new APIs offer us the chance to offer GraphQL as a default option for users of EmberData. While we are not pivoting away from a strong story for JSON:API users, we want both stories to be equally strong (and even encourage mixing the two, they complement each other well).

## Action Items

This is a non-comprehensive list that we expect to grow over the next several months as we determine the exact mechanics of achieving the above.

**Deprecations**
- [-] deprecate Legacy Imports https://github.com/emberjs/rfcs/pull/743
- [-] deprecate ajax for fetch https://github.com/emberjs/rfcs/pull/790
- [x] modernize PromiseManyArray https://github.com/emberjs/rfcs/pull/745
- [x] simplify schema type signatures https://github.com/emberjs/rfcs/pull/794/
- [x] deprecate helper functions https://github.com/emberjs/rfcs/pull/742
- [x] deprecate accessing static fields on model prior to lookup https://github.com/emberjs/rfcs/pull/741
- [-] deprecate non-strict types https://github.com/emberjs/rfcs/pull/740
- [x] deprecate non-strict relationships https://github.com/emberjs/rfcs/pull/739
- [x] deprecate model reopen https://github.com/emberjs/rfcs/pull/738
- [x] deprecate extending native error https://github.com/emberjs/rfcs/issues/402
- [x] deprecate setting via proxy https://github.com/emberjs/rfcs/issues/397
- [-] deprecate missing data for sync relationships being considered empty https://github.com/emberjs/rfcs/issues/360
- [-] deprecate store normalization/serialization APIs https://github.com/emberjs/rfcs/issues/357
- [x] deprecate Store/Adapter/Serializer/Model EmberObject APIs
- [x] deprecate classic chains, A(), and array-like
- [-] deprecate serializer.extractErrors
- [-] deprecate store.modelFor
- [-] deprecate Model init / property access during init
- [x] deprecate store's promise-proxy return values https://github.com/emberjs/rfcs/pull/846
- [x] deprecate array-like APIs https://github.com/emberjs/rfcs/pull/846
- [-] deprecate setting unknown args on Model
- [-] deprecate RecordArrays being passed to adapters
- [-] deprecate null values for collection relationships
- [-] deprecate accessing values on records post-destroy

**Refactoring**
- [x] complete removal of InternalModel
- [x] move existing coalescing logic into the adapters
- [-] return promises from unloadRecord/unloadAll https://github.com/emberjs/rfcs/issues/385
- [x] replace RSVP.Promise with Native Promise usage

**Documentation**
- [x] document store APIs for custom model classes
- [x] document store APIs for RecordData
- [x] document store APIs for schema
- [x] document RecordData
- [x] document store APIs for Request notifications
- [x] write an overview for ember-data https://github.com/ember-learn/ember-api-docs/pull/761
- [x] README for store package
- [x] README for request package
- [x] README for adapter package
- [x] README for serializer package
- [x] README for debug package
- [x] README for json:api cache package
- [x] README for graph package
- [x] README for model package
- [x] overview for store package
- [x] overview for request package
- [x] overview for adapter package
- [x] overview for serializer package
- [x] overview for debug package
- [x] overview for json:api cache package
- [x] overview for graph package
- [x] overview for model package

**Features**
- [x] Polymorphic support without Mixins https://github.com/emberjs/rfcs/pull/793
- [x] Implement RecordDataV2 https://github.com/emberjs/rfcs/pull/461
- Implement Cache V2.1 https://github.com/emberjs/rfcs/pull/854
  - [x] Singleton Cache
  - [x] Store.createCache
  - [x] Store.cache
  - [x] Store.instantiateRecord (updated)
  - [x] Store.notifications
  - [x] Resource Added/Removed Notifications
  - [x] Cache.upsert
  - [x] Cache.patch
  - [x] Cache.mutate
  - [-] Cache.fork
  - [-] Cache.merge
  - [-] Cache.diff
  - [-] Cache.dump
  - [-] Cache.hydrate
  - Cache.peekRequest
    - [x] Data Documents
    - [x] Error Documents
    - [x] Meta Documents
  - Cache.peek
    - [x] Data Documents
    - [x] Resource Blobs
    - [x] Error Documents
    - [x] Meta Documents
  - Cache.put
    - [x] Data Documents
    - [x] Error Documents
    - [x] Meta Documents
- [-] introduce operations for delete and relationship updates https://github.com/emberjs/rfcs/issues/355 https://github.com/emberjs/rfcs/pull/458
- [x] Document|Collection (Query) Cache
- [x] RequestManager
  - [-] replacement middleware default for JSON:API and REST
  - [x] refactor existing request methods to utilize request-manager
  - [x] Store.request
  - [x] Store.requestManager
  - [x] `ember-data` auto-configures RequestManager
  - [-] request-utils
  - [x] Store.lifetimes
  
 **Request Manager Support**
 
- [x] improve enforcement of immutability in dev
- [x] additional test coverage for Store.request, errors thrown during request
- [x] additional test coverage for cache.peek / cache.put / cache.peekRequest
- [x] update docs for RequestManager.useCache
- [-] Fetch Handler | handle returned non-JSON
- [x] Fetch Handler | handle abort
- implement new `Collection` mode to properly reflect `ResourceDocument`
  - [x] Document Wrapper Class
    - [x] Collection Support
    - [x] Pagination Support
- [x] ensure single-resource-document requests are notified when `data` changes
- [-] determine Cache "mutability" of `CollectionResourceDocument`
- [-] add tests for enhancing `context.request` in RequestManager before calling next

Deprecation Guides

- [x] DEPRECATE_RSVP_PROMISE
- [x] DEPRECATE_SAVE_PROMISE_ACCESS
- [x] DEPRECATE_SNAPSHOT_MODEL_CLASS_ACCESS
- [x] DEPRECATE_STORE_FIND
- [x] DEPRECATE_HAS_RECORD
- [x] DEPRECATE_STRING_ARG_SCHEMAS
- [x] DEPRECATE_JSON_API_FALLBACK
- [x] DEPRECATE_MODEL_REOPEN
- [x] DEPRECATE_EARLY_STATIC
- [x] DEPRECATE_HELPERS
- [x] DEPRECATE_PROMISE_MANY_ARRAY_BEHAVIORS
- [x] DEPRECATE_V1CACHE_STORE_APIS
- [x] DEPRECATE_RELATIONSHIPS_WITHOUT_TYPE
- [x] DEPRECATE_RELATIONSHIPS_WITHOUT_ASYNC
- [x] DEPRECATE_RELATIONSHIPS_WITHOUT_INVERSE
- [x] DEPRECATE_V1_RECORD_DATA
- [x] DEPRECATE_A_USAGE
- [x] DEPRECATE_PROMISE_PROXIES
- [x] DEPRECATE_ARRAY_LIKE
- [x] DEPRECATE_COMPUTED_CHAINS
- [x] DEPRECATE_NON_EXPLICIT_POLYMORPHISM
- [x] DEPRECATE_INSTANTIATE_RECORD_ARGS
- [x] DEPRECATE_CREATE_RECORD_DATA_FOR_HOOK

Other Notes:

- [x] clarify abort signal handling in the RFC / handler documentation
- [x] clarify intended Error handling mechanisms/propagation
- [x] update RFC for RequestManager.useCache
- [x] update RFC for `lifetimes.isSoftExpired/isHardExpired`
- [x] update RFC for rename `data` => `content`
  
**Moved to 6.0**
- [-] Model-less EmberData
  - [-] Collection Primitive
  - [-] Paginated Collection
  - [-] Support Nested Relationship Keys
  - [-] ProxyModel supports nested attributes
  - [-] schema decorators and schema model for transitioning users

**Comment by runspired** [comment]
A potential thing here is to deprecate unstable identifier API signatures. As we've worked through actual implementations this overhead has turned out to be largely unnecessary, at times it even introduces potential error situations (caches doing operations in an unsupported order), and the ergonomics of working with the stable identifier are far better for both caches and the store anyway: especially with typescript.

**Comment by sandstrom** [comment]
Great roadmap! 🏅 

Removing EmberObject, observers, etc. sound like a very good idea.

## Thoughts on snapshots and forks

I don't want to pile on additional things to the roadmap. Just mentioning this as more of a background feedback and general thought, rather than something that I think should necessarily be in 5.0.

The ability to make snapshots (or forks) of model state (attributes + relationships) and then roll back is something that would make working with forms easier. Basically, a lot of forms have an edit/cancel button, and if you attach a model to the form, make some changes and then press cancel it's nice to have an easy way of rolling it all back.

I know there are snapshots internally, exposed to serializers. But they aren't accessible "from the outside".

There are also some prior addons:
- https://github.com/danielspaniel/ember-data-change-tracker
- https://github.com/dylanmensaert/ember-rollback-relationships

Today we are using [Ember Changeset](https://github.com/poteto/ember-changeset) to handle forms, and are using a bunch of workarounds to buffer changes to relationships. I've also mentioned this to @snewcomer in https://github.com/poteto/ember-changeset/issues/551.

Would the ideal place for this be Ember Data or Ember Changeset? I don't know. But maybe it's a primitive that should reside in Ember Data, so I'm mentioning it here.

Also, maybe having the use-case of snapshots/forks in mind early in the planning, may make it easier to add them in the future, even if it isn't in 5.0.

Happy to clarify and explain more about our use-case. There is also a code example in the [aforementioned issue](https://github.com/poteto/ember-changeset/issues/551).

**Comment by runspired** [comment]
@sandstrom internally it's actually possible at this point to rollback relationships. This capability is not exposed to users because historically rollback did not include that.

With the V2 Cache, store-forking becomes something we can (relatively easily) add. Store forking is our intended route for handling all the various encapsulation scenarios where you want to be able to group a set of changes and/or quickly discard a set of records and/or a set of changes.

**Comment by sandstrom** [comment]
Sounds awesome @runspired!

**Comment by runspired** [comment]
Added `deprecate accessing record props after destroy`. This behavior was already difficult to support and partially broken by changes in 4.4 and again 4.5. The singleton cache exposed this problem again in a sever way in #8122, the result was adding a secret second cache to make access during teardown "mostly safe" for one browser lifecycle tick.

```ts
unloadRecord(identifier: StableRecordIdentifier): void {
    const cached = this.#peek(identifier);
    const storeWrapper = this.#storeWrapper;
    graphFor(storeWrapper).unload(identifier);

    // effectively clearing these is ensuring that
    // we report as `isEmpty` during teardown.
    cached.localAttrs = null;
    cached.remoteAttrs = null;
    cached.inflightAttrs = null;

    let relatedIdentifiers = _allRelatedRecordDatas(storeWrapper, identifier);
    if (areAllModelsUnloaded(storeWrapper, relatedIdentifiers)) {
      for (let i = 0; i < relatedIdentifiers.length; ++i) {
        let identifier = relatedIdentifiers[i];
        storeWrapper.disconnectRecord(identifier);
      }
    }

    this.#cache.delete(identifier);
    this.#destroyedCache.set(identifier, cached);

    /*
     * The destroy cache is a hack to prevent applications
     * from blowing up during teardown. Accessing state
     * on a destroyed record is not safe, but historically
     * was possible due to a combination of teardown timing
     * and retention of a RecordData instance directly on the
     * record itself.
     *
     * Once we have deprecated accessing state on a destroyed
     * instance we may remove this.
     */
    if (this.#destroyedCache.size === 1) {
      schedule('destroy', () => {
        setTimeout(() => {
          this.#destroyedCache.clear();
        });
      });
    }
  }
  ```

**Comment by runspired** [comment]
And we're done 🥳

**Comment by sandstrom** [comment]
@runspired Awesome! 💯 

Happy Easter! 🐰🥚
